File: src/main/java/leetCode/t876/java/Solution.java
package leetCode.t876.java;

class Solution {
    public ListNode middleNode(ListNode head) {
        ListNode middle = head;
        ListNode end = head;

        while (end != null && end.next != null) {
            middle = middle.next;
            end = end.next.next;
        }
        return middle;
    }
}

File: src/main/java/leetCode/t876/kotlin/Solution.kt
package leetCode.t876.kotlin

class Solution {
    fun middleNode(head: ListNode?): ListNode? {
        var middle = head
        var end = head

        while (end?.next != null) {
            middle = middle?.next
            end = end.next?.next
        }
        return middle
    }
}

class ListNode(var `val`: Int) {
    var next: ListNode? = null
}